src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -1692,126 +1692,134 @@ private static Guid GenerateGuidFromName(string name)
             return new Guid(bytes);
         }
 
-        private static unsafe object? DecodeObject(Type dataType, ref EventSource.EventData* data)
+        private static unsafe object? DecodeObject(Type dataType, ref EventData* data)
         {
             // TODO FIX : We use reflection which in turn uses EventSource, right now we carefully avoid
             // the recursion, but can we do this in a robust way?
 
             IntPtr dataPointer = data->DataPointer;
-            // advance to next EventData in array
-            ++data;
 
-            Again:
-            if (dataType == typeof(IntPtr))
+            switch (RuntimeTypeHandle.GetCorElementType((RuntimeType)dataType))
```

Comment:
I expect that you will see build breaks here.  The portable way to do this is using `Type.GetTypeCode`